pulsar/src/main/java/com/hazelcast/jet/contrib/pulsar/PulsarSources.java

```diff
@@ -0,0 +1,374 @@
+     * @param projectionFn   converts a Pulsar message to an emitted item.
+     * @param <M>            the type of the message read by {@code PulsarConsumer}
+     * @param <T>            the type of data emitted from {@code StreamSource}
+     * @return a stream source to use in {@link Pipeline#readFrom}
+     */
+
+    public static <M, T> StreamSource<T> pulsarReader(
+            @Nonnull String topic,
+            @Nonnull Map<String, Object> readerConfig,
+            @Nonnull SupplierEx<PulsarClient> connectionSupplier,
+            @Nonnull SupplierEx<Schema<M>> schemaSupplier,
+            @Nonnull FunctionEx<Message<M>, T> projectionFn
+    ) {
+        checkSerializable(connectionSupplier, "connectionSupplier");
+        return SourceBuilder.timestampedStream("pulsar-ft-stream-source", ctx -> new ReaderContext(
+                ctx.logger(), connectionSupplier.get(), topic, readerConfig, schemaSupplier, projectionFn))
+                .<T>fillBufferFn(ReaderContext::fillBuffer)
+                .createSnapshotFn(ReaderContext::createSnapshot)
+                .restoreSnapshotFn(ReaderContext::restoreSnapshot)
+                .destroyFn(ReaderContext::destroy)
+                .build();
+    }
+
+
+    /**
+     * A context for the stream source of Apache Pulsar
+     *
+     * @param <M> the type of the value of message read by {@code PulsarConsumer}
+     * @param <T> the type of the emitted item after projection.
+     */
+    private static final class ConsumerContext<M, T> {
+        private static final int MAX_NUM_MESSAGES = 512;
+        private static final int TIMEOUT_IN_MS = 1000;
+        private static final int MAX_ACK_RETRIES = 10;
+
+        private final ILogger logger;
+        private final PulsarClient client;
+        private final Consumer<M> consumer;
+        private final FunctionEx<Message<M>, T> projectionFn;
+
+        private ConsumerContext(
+                @Nonnull ILogger logger,
+                @Nonnull PulsarClient client,
+                @Nonnull List<String> topics,
+                @Nonnull Map<String, Object> consumerConfig,
+                @Nonnull SupplierEx<Schema<M>> schemaSupplier,
+                @Nonnull FunctionEx<Message<M>, T> projectionFn
+        ) throws PulsarClientException {
+            checkSerializable(schemaSupplier, "schemaSupplier");
+            checkSerializable(projectionFn, "projectionFn");
+            this.logger = logger;
+            this.projectionFn = projectionFn;
+            this.client = client;
+            this.consumer = client.newConsumer(schemaSupplier.get())
+                                  .topics(topics)
+                                  .loadConf(consumerConfig)
+                                  .subscriptionType(SubscriptionType.Shared)
+                                  .subscriptionInitialPosition(SubscriptionInitialPosition.Earliest)
+                                  .batchReceivePolicy(BatchReceivePolicy.builder()
```

Comment:
Let's expose batchReceivePolicySupplier. I bet the 1st user would complain about this hardcoded setting :-)